File: 20. Valid Parentheses.java
public class Solution {
    public boolean isValid(String s) {
        Stack stk = new Stack();
        
        for (int i = 0; i < s.length(); i++) {
            char a = s.charAt(i);
            
            if (a == '(' || a == '[' || a == '{') {
                stk.push(a);
            } else if (!stk.empty()) {
                char b = (char) stk.peek();
                
                if (a == ')' && b == '(') {
                    stk.pop();
                } else if (a == ']' && b == '[') {
                    stk.pop();
                } else if (a == '}' && b == '{') {
                    stk.pop();
                } else {
                    return false;
                }
            } else {
                return false;
            }
        }
        return stk.empty();
    }
}

File: 20. Valid Parentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        
        for p in s:
            if p == "(" or p == "[" or p == "{":
                stack.append(p)
            elif stack:
                curr = stack.pop()
                if p == ")" and curr != "(":
                    return False
                elif p == "]" and curr != "[":
                    return False
                elif p == "}" and curr != "{":
                    return False
            else:
                return False
        
        if not stack:
            return True
        else:
            return False
